# Machine Learning and Data Mining: Graded Lab 01
---------------

#### Grading
For this graded lab you can get a total of 15 points. These 15 points count 10% of your final grade for the course.

#### Deadline
You have 20 days to finish the lab. You must submit the graded lab the evening before the lab-session. 

- **If you have lab on Friday: Hand-in on 11.11.21, 11.59 p.m.**
- **If you have lab on Saturday: Hand-in on 12.11.21, 11.59 p.m.**

#### Submission
You have to solve the labs in groups of **two** students. If the class has an uneven number of students, there can be one group of three students. For this group, however, task 3 is graded more strictly and extra effort is expected.

Submit your GradedLab01.pynb file renamed to **LastnameStudent01_LastnameStudent02.ipynb** in moodle.   
Please submit a runnable python notebook file **and** an exported HTML with all cells executed (the results must be visible in the HTML).
All other submissions will be rejected and graded with 0 points.

#### Presentation Exercise 3
The students are completely free to decide how to solve exercise 3. If the exercise is not solved in this notebook, a detailed description with screenshots must be included. In any case, task 3 must be presented to the lecturer during class. Thereby, a live demonstration is expected and not a presentation with PowerPoint. This demonstration should take about 5 minutes.

In [2]:
#Wichtiges:
#-lauffähiges Python notebook und mit exportierter HTMLversion -> Sonst Note1
#-Projektdatei muss so heissen: PioLogin01_Arslan02.ipynb
#-Aufgabe3 muss präsentiert werden, ca. 5min live oder über MS-Team. Kein Powerpoint. 

---

## Exercise 1: Implement a content-based recommender for the data in the movies.csv file [4 points].    
The movies.csv file contains a dataset extracted from IMDB with several attributes. For example, the title, plot, and the language of the movies.

##### __(a) The user Calvin hasn't watched any movies yet. Print the top 5 movies based on the rating attribute, as recommendations for him [0.5 points].__ 

In [51]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

df = pd.read_csv('movies.csv')
df

id                                              title  \
0   2897196                         The Art of the Silent Fist   
1   1329338                                 Dla ciebie i ognia   
2   7065882                                         Red Dragon   
3   1615477  The Half-Life Horror from Hell or: Irradiated ...   
4   2316564                                       Nenek Gayung   
5    174856                                      The Hurricane   
6    244847                                    Dark Blue World   
7    104927                                           Mr. Bond   
8   1562870                     Prince Charles' Other Mistress   
9    967060                             Game, Set... Muuurder?   
10   148537                                        Not Ghoulty   
11   260775                                           Calle 54   
12  5280788  Forbidden Archeology: The Hidden History of th...   
13   502413                                             Frenzy   
14   371636                              Dil Ne Phir Yaad Kiya   
15  1809195                              Norwegian Cruise Line   
16  2130022                                        Enchantress   
17   419815                                              Holly   
18  2187105                                 The Backward Class   
19  6610020                         Brainstorm: Between Shores   
20   642381                                 Sue Casa, His Casa   
21   288320                                        The Wounded   
22  2932788                                          Homerland   
23  2912226                                Unleash the Kraken!   
24   559276                                      Teacher's Pet   
25  3295276                The Secret Sex Life of a Single Mom   
26  4138326                                            Fallout   
27   660187                                      Freaky Friday   
28  1634540  Rescue Men: The Story of the Pea Island Lifesa...   
29   110912                                       Pulp Fiction   
30   105236                                     Reservoir Dogs   
31   116367                                From Dusk Till Dawn   
32  1028528                                        Death Proof   
33  3460252                                  The Hateful Eight   
34   361748                               Inglourious Basterds   
35   119396                                       Jackie Brown   
36   111161                           The Shawshank Redemption   
37    68646                                      The Godfather   
38    71562                             The Godfather: Part II   
39   468569                                    The Dark Knight   
40  1502397                                  Bad Boys for Life   
41   380268                                    Cocaine Cowboys   
42  7056864                                  Cocaine Godmother   
43  1176726                                  Cocaine Cowboys 2   

                                            languages  \
0                                         ['English']   
1                                          ['Polish']   
2                                        ['Japanese']   
3                                         ['English']   
4                                      ['Indonesian']   
5                                         ['English']   
6            ['Czech', 'German', 'English', 'Slovak']   
7                                           ['Hindi']   
8                                         ['English']   
9                                         ['English']   
10                                        ['English']   
11                   ['French', 'Spanish', 'English']   
12                                        ['English']   
13                                        ['English']   
14                      ['Hindi', 'Urdu', 'Bhojpuri']   
15                                        ['English']   
16                                        ['English']   
17          

In [52]:
dfSortedByRating = df.sort_values(by=['rating'])
dfTop = dfSortedByRating.tail(5)
dfTop

id                     title  \
29  110912              Pulp Fiction   
38   71562    The Godfather: Part II   
39  468569           The Dark Knight   
37   68646             The Godfather   
36  111161  The Shawshank Redemption   

                                            languages  \
29                   ['English', 'Spanish', 'French']   
38  ['English', 'Italian', 'Spanish', 'Latin', 'Si...   
39                            ['English', 'Mandarin']   
37                    ['English', 'Italian', 'Latin']   
36                                        ['English']   

                                      genres  rating              director  \
29                        ['Crime', 'Drama']     8.9     Quentin Tarantino   
38                        ['Crime', 'Drama']     9.0  Francis Ford Coppola   
39  ['Action', 'Crime', 'Drama', 'Thriller']     9.0     Christopher Nolan   
37                        ['Crime', 'Drama']     9.2  Francis Ford Coppola   
36                                 ['Drama']     9.3        Frank Darabont   

                                                 plot  
29  Jules Winnfield (Samuel L. Jackson) and Vincen...  
38  The continuing saga of the Corleone crime fami...  
39  Set within a year after the events of Batman B...  
37  The Godfather "Don" Vito Corleone is the head ...  
36  Chronicles the experiences of a formerly succe...

##### __(b) The user Calvin hasn't still watched any movies yet. We  just know that he likes english speaking movies. Print the top 5 movies based on the rating attribute, which have only the language "English" in the languages attribute, as recommendations for him [0.5 points].__ 

In [56]:
dfFiltered = df[df['languages'] == '[\'English\']']
dfFiltered

id                                              title    languages  \
0   2897196                         The Art of the Silent Fist  ['English']   
3   1615477  The Half-Life Horror from Hell or: Irradiated ...  ['English']   
5    174856                                      The Hurricane  ['English']   
8   1562870                     Prince Charles' Other Mistress  ['English']   
9    967060                             Game, Set... Muuurder?  ['English']   
10   148537                                        Not Ghoulty  ['English']   
12  5280788  Forbidden Archeology: The Hidden History of th...  ['English']   
13   502413                                             Frenzy  ['English']   
15  1809195                              Norwegian Cruise Line  ['English']   
16  2130022                                        Enchantress  ['English']   
18  2187105                                 The Backward Class  ['English']   
20   642381                                 Sue Casa, His Casa  ['English']   
21   288320                                        The Wounded  ['English']   
22  2932788                                          Homerland  ['English']   
23  2912226                                Unleash the Kraken!  ['English']   
24   559276                                      Teacher's Pet  ['English']   
25  3295276                The Secret Sex Life of a Single Mom  ['English']   
26  4138326                                            Fallout  ['English']   
27   660187                                      Freaky Friday  ['English']   
28  1634540  Rescue Men: The Story of the Pea Island Lifesa...  ['English']   
30   105236                                     Reservoir Dogs  ['English']   
32  1028528                                        Death Proof  ['English']   
35   119396                                       Jackie Brown  ['English']   
36   111161                           The Shawshank Redemption  ['English']   
41   380268                                    Cocaine Cowboys  ['English']   
42  7056864                                  Cocaine Godmother  ['English']   
43  1176726                                  Cocaine Cowboys 2  ['English']   

                                               genres  rating  \
0   ['Animation', 'Action', 'Adventure', 'Comedy',...     8.3   
3                                ['Comedy', 'Horror']     7.2   
5                     ['Biography', 'Drama', 'Sport']     7.6   
8                                     ['Documentary']     7.1   
9                      ['Comedy', 'Crime', 'Mystery']     7.7   
10  ['Animation', 'Short', 'Comedy', 'Family', 'Fa...     6.4   
12                                    ['Documentary']     4.7   
13                                         ['Comedy']     8.4   
15                                     ['Reality-TV']     6.2   
16                                        ['Fantasy']     5.0   
18                           ['Documentary', 'Drama']     8.6   
20                                         ['Comedy']     7.8   
21                                ['Action', 'Drama']     6.9   
22                            ['Animation', 'Comedy']     6.8   
23                     ['Comedy', 'Drama', 'Fantasy']     8.6   
24                      ['Crime', 'Drama', 'Mystery']     7.2   
25                    ['Drama', 'Mystery', 'Romance']     4.7   
26         ['Action', 'Adventure', 'Drama', 'Sci-Fi']     8.9   
27                              ['Comedy', 'Romance']     7.2   
28                                    ['Documentary']     8.2   
30                     ['Crime', 'Drama', 'Thriller']     8.3   
32                ['Action', 'Adventure', 'Thriller']     7.0   
35                     ['Crime', 'Drama', 'Thriller']     7.5   
36                                          ['Drama']     9.3   
41                ['Documentary', 'Crime', 'History']     7.8   
42                    ['Biography', 'Crime', 'Drama']     5.6   
43                ['Documentary', 'Crime'

In [58]:
dfSorted = dfFiltered.sort_values(by=['rating'])
dfTopEnglish = dfSorted.tail(5)
dfTopEnglish

id                     title    languages  \
13   502413                    Frenzy  ['English']   
23  2912226       Unleash the Kraken!  ['English']   
18  2187105        The Backward Class  ['English']   
26  4138326                   Fallout  ['English']   
36   111161  The Shawshank Redemption  ['English']   

                                        genres  rating         director  \
13                                  ['Comedy']     8.4    Richard Boden   
23              ['Comedy', 'Drama', 'Fantasy']     8.6       Mike Smith   
18                    ['Documentary', 'Drama']     8.6  Madeleine Grant   
26  ['Action', 'Adventure', 'Drama', 'Sci-Fi']     8.9   Stephen Surjik   
36                                   ['Drama']     9.3   Frank Darabont   

                                                 plot  
13  The new kitchen is now complete but Bill is to...  
23  Just as the deadbeat father of the Johnson bro...  
18  The first class of Dalit ('untouchable') caste...  
26  After the nuclear explosion, General Wade Eili...  
36  Chronicles the experiences of a formerly succe...

##### __(c) Still Calvin hasn't still watched any movies yet. However, we noticed that he took more time on the recommendations with the genres "Drama" and "Crime", therefore we assume that he is interested in this genre combination. Print the top 3 movies based on the rating attribute, which have at least the genres "Drama" and "Crime" as recommendations for him [0.5 points].__ 

In [62]:
dfSortedCrimeDrama = df[(df['languages'] == '[\'English\']') & (df['genres'].str.contains('Drama' or 'Crime'))]
dfSortedCrimeDrama

id                                title    languages  \
5    174856                        The Hurricane  ['English']   
18  2187105                   The Backward Class  ['English']   
21   288320                          The Wounded  ['English']   
23  2912226                  Unleash the Kraken!  ['English']   
24   559276                        Teacher's Pet  ['English']   
25  3295276  The Secret Sex Life of a Single Mom  ['English']   
26  4138326                              Fallout  ['English']   
30   105236                       Reservoir Dogs  ['English']   
35   119396                         Jackie Brown  ['English']   
36   111161             The Shawshank Redemption  ['English']   
42  7056864                    Cocaine Godmother  ['English']   

                                        genres  rating           director  \
5              ['Biography', 'Drama', 'Sport']     7.6     Norman Jewison   
18                    ['Documentary', 'Drama']     8.6    Madeleine Grant   
21                         ['Action', 'Drama']     6.9   John Azpilicueta   
23              ['Comedy', 'Drama', 'Fantasy']     8.6         Mike Smith   
24               ['Crime', 'Drama', 'Mystery']     7.2   Vincent McEveety   
25             ['Drama', 'Mystery', 'Romance']     4.7      John L'Ecuyer   
26  ['Action', 'Adventure', 'Drama', 'Sci-Fi']     8.9     Stephen Surjik   
30              ['Crime', 'Drama', 'Thriller']     8.3  Quentin Tarantino   
35              ['Crime', 'Drama', 'Thriller']     7.5  Quentin Tarantino   
36                                   ['Drama']     9.3     Frank Darabont   
42             ['Biography', 'Crime', 'Drama']     5.6  Guillermo Navarro   

                                                 plot  
5   This film tells the story of Rubin "Hurricane"...  
18  The first class of Dalit ('untouchable') caste...  
21  Los Angeles, the city of angels where six murd...  
23  Just as the deadbeat father of the Johnson bro...  
24  Mark and Steve go fishing; when they return th...  
25  Marcus and Mike have to confront new issues (c...  
26  After the nuclear explosion, General Wade Eili...  
30  Six criminals, who are strangers to each other...  
35  The middle-aged stewardess Jackie Brown smuggl...  
36  Chronicles the experiences of a formerly succe...  
42  The life of drug lord Griselda Blanco, a pione...

In [65]:
dfSortedCD = dfSortedCrimeDrama.sort_values(by=['rating'])
dfTopCD = dfSortedCD.tail(3)
dfTopCD

id                     title    languages  \
23  2912226       Unleash the Kraken!  ['English']   
26  4138326                   Fallout  ['English']   
36   111161  The Shawshank Redemption  ['English']   

                                        genres  rating        director  \
23              ['Comedy', 'Drama', 'Fantasy']     8.6      Mike Smith   
26  ['Action', 'Adventure', 'Drama', 'Sci-Fi']     8.9  Stephen Surjik   
36                                   ['Drama']     9.3  Frank Darabont   

                                                 plot  
23  Just as the deadbeat father of the Johnson bro...  
26  After the nuclear explosion, General Wade Eili...  
36  Chronicles the experiences of a formerly succe...

##### __(d) Now, Calvin has watched the movie with the ID 110912. Print the top 3 movies based on the rating attribute, which have the same director as the watched movie with the ID 110912 as recommendations for him. The already watched movie should not be within the recommendations [0.5 points].__ 

In [66]:
GetMovie = df[df['id'] == 110912]
GetMovie
## What are the rating attributes? just the rating or also the languages and genres? 

id         title                         languages  \
29  110912  Pulp Fiction  ['English', 'Spanish', 'French']   

                genres  rating           director  \
29  ['Crime', 'Drama']     8.9  Quentin Tarantino   

                                                 plot  
29  Jules Winnfield (Samuel L. Jackson) and Vincen...

##### __(d) Now, Calvin has watched the movie with the ID 1502397 and we want to recommend movies with a similar plot. Remove all stopwords from the plot attribute and split the content of the plot attribute into terms. Then calculate the term similarity between the plot of the watched movie and all others. Use data pre-processing methods wherever needed. Print the top 3 movies based on the calculated term similarity and the rating of the movie.  [2 points].__ 

---

## Exercise 2: Implement Collaborative Filtering Recommenders for the data in the ratings.csv file [6 points].
##### __The ratings.csv file contains a dataset extracted from IMDB with ratings for movies from several users. Implement a user-based and an item-based collaborative filtering recommender. Use data pre-processing methods wherever needed. Print the top 3 recommended movies for each user in the dataset for the user-based and the item-based version. Don't recommend any movies the user has already watched.__ 

In [1]:
users = ['not implemented']

def getRecommendations(user):
    return 'not implemented.'

for user in users:
    print('The user-based recommendations for ' + user + ' are: ' + 
          str(getRecommendations(user)))
    print('The item-based recommendations for ' + user + ' are: ' + 
          str(getRecommendations(user)))

The user-based recommendations for not implemented are: not implemented.
The item-based recommendations for not implemented are: not implemented.


---

## Exercise 3: Extend the Collaborative Filtering Recommender [5 points].

You are free to choose which additional task you want to do. Ideas could be:
- Implement your own UI for the Recommendation System.
- Create a blogpost on the internet describing your approach and the system (Note: Do not just upload the notebook but create a text with references to the literature and appropriate plots).
- Test alternative approaches to create recommendations and compare them.
- **... your own idea ...**

If you solve this task in this notebook, just describe your idea in a few sentences. If you do not solve it in the notebook, please provide a more detailed description and screenshots if appropriate.

This exercise must be presented during the lab session. The task is graded based on innovation, effort, result, ...

### Describe your idea here:

.....

